src/lib.rs

```diff
@@ -76,28 +86,372 @@ impl ULinkZenoh {
+                )));
+                return;
+            };
+            // Create UPayload
+            let u_payload = match query.value() {
+                Some(value) => {
+                    let Ok(encoding) = value.encoding.suffix().parse::<i32>() else {
+                        listener(Err(UStatus::fail_with_code(
+                            UCode::Internal,
+                            "Unable to get payload encoding",
+                        )));
+                        return;
+                    };
+                    UPayload {
+                        length: Some(0),
+                        format: encoding,
+                        data: Some(Data::Value(value.payload.contiguous().to_vec())),
+                    }
+                }
+                None => UPayload {
+                    length: Some(0),
+                    format: UPayloadFormat::UpayloadFormatUnspecified as i32,
+                    data: None,
+                },
+            };
+            // Create UMessage
+            let msg = UMessage {
+                source: Some(method.clone()),
+                attributes: Some(u_attribute.clone()),
+                payload: Some(u_payload),
+            };
+            if let Some(reqid) = u_attribute.reqid {
+                query_map
+                    .lock()
+                    .unwrap()
+                    .insert(ULinkZenoh::uuid_to_string(&reqid), query);
+            } else {
+                listener(Err(UStatus::fail_with_code(
+                    UCode::Internal,
+                    "The request is without reqid in UAttributes",
+                )));
+                return;
+            }
+            listener(Ok(msg));
+        };
+        if let Ok(queryable) = self
+            .session
+            .declare_queryable(&zenoh_key)
+            .callback_mut(callback)
+            .res()
+            .await
+        {
+            self.queryable_map
+                .lock()
+                .unwrap()
+                .insert(hashmap_key.clone(), queryable);
+        } else {
+            return Err(UStatus::fail_with_code(
+                UCode::Internal,
+                "Unable to register callback with Zenoh",
+            ));
+        }
+
+        Ok(hashmap_key)
+    }
+    async fn unregister_rpc_listener(&self, method: UUri, listener: &str) -> Result<(), UStatus> {
+        // Do the validation
+        // TODO: Whether we should check URI is resolved or not
+        //       https://github.com/eclipse-uprotocol/uprotocol-spec/issues/42#issuecomment-1882273981
+        if UriValidator::validate(&method).is_err() {
+            return Err(UStatus::fail_with_code(
+                UCode::InvalidArgument,
+                "Invalid topic",
+            ));
+        }
+        // TODO: Check whether we still need method or not (Compare method with listener?)
+
+        if !self.queryable_map.lock().unwrap().contains_key(listener) {
```

Comment:
Alternatively, you can directly perform the `remove()` operation and check if the returned value `is_none()`.

Reply:
Cool, that makes the code simpler.